### ISSUE TITLE
Don't know how to coerce an object of class 'sheets_meta/list' into a 'sheets_id'

### ISSUE THREAD

**Issue by SridharJagannathan** [tidyverse/googlesheets4#74]
I have been trying to use the `googlesheets4` using sheetid like below

```r
ss <- googlesheets4::sheets_get(id = xxxyyyzzz)
googlesheets4::read_sheet(ss, sheet = 1)
```
This returns the error
```r
Error: Don't know how to coerce an object of class 'sheets_meta/list' into a 'sheets_id'
```

The tested version was from `CRAN 0.1.0`
However the same code above runs fine on the development version in `Github 0.1.0.9000`
On further investigation I found that the class of `ss` above is `"sheets_meta" "list"`  in the CRAN version however in the GitHub version it is `"sheets_Spreadsheet" "list"`.
This is causing an error inside the `get_cells` function in the line `ssid <- as_sheets_id(ss)`
Could you please let me know when would the develop version be put in the `CRAN` with this fix.

**Comment by jennybc** [comment]
What does `get_team_sheet()` do and return? That is your own function.

Re: "it works in dev", yeah I have created more methods for making more outputs usable as input for other functions.

But in general you should be handing a Sheet ID around. So above you probably should do something like:

``` r
googlesheets4::read_sheet(as_sheets_id(ss), sheet = 1)
```

**Comment by SridharJagannathan** [comment]
Hi @jennybc thanks for this, I have updated the example, in essence the `get_team_sheet()` is a wrapper around `googlesheets4::sheets_get`. Would you recommend something like 
```r
ss <- googlesheets4::sheets_get(id = xxxyyyzzz)
googlesheets4::read_sheet(as_sheets_id(ss), sheet = 1)
```
Which would mean that the `read_sheet` should always be supplied with an object of class `"sheets_id"`. I will check this and update you if this works.

**Comment by jennybc** [comment]
As you noted, your original idea *does* work in the dev version. I am starting to put `as_sheets_id()` into the functions that need to operate on a Sheet, so that all of us don't trip over this repeatedly. It's annoying.

That being said, the googlesheets4 package does really regard the Sheet ID as the main/only identifier. And that's the only thing that really cannot change: all the other meta-data returned by `sheets_get()` is changeable (file name, constituent worksheets, etc.).

So if you're trying to write a script in a robust, programming-oriented style, that suggests using Sheet ID as the identifier.

**Comment by SridharJagannathan** [comment]
Hi @jennybc I have tried out your suggestion like below (however it still seems to give the error?)

```r
ss <- googlesheets4::sheets_get(id = xxxyyyzzz)
googlesheets4::read_sheet(googlesheets4::as_sheets_id(ss), sheet = 1)

Error: Don't know how to coerce an object of class 'sheets_meta/list' into a 'sheets_id'

```

This time the error is coming from the function `googlesheets4::as_sheets_id`

**Comment by SridharJagannathan** [comment]
I think the main problem here is
```R
class(ss)
[1] "sheets_meta" "list" 
```
There is no method in `googlesheets4::as_sheets_id` to handle objects of type `sheets_meta` and it goes to the default method which is `as_sheets_id.default` which returns the error. In the dev version the 

```R
class(ss)
[1] "googlesheets4_spreadsheet" "list" 
```
Here there is a method to handle objects of type `googlesheets4_spreadsheet` in `googlesheets4::as_sheets_id`

I think this is a bug to be fixed in the `cran` version and not a programming practise issue.

**Comment by jennybc** [comment]
Ok then use ss$spreadsheet_id for now. As we’ve noted, this is already fixed in dev.

**Comment by kleinschmidt** [comment]
I ran into it just following the directions from the docs page (no wrappers or anything), by trying to pass the output of `sheets_get` to `sheets_sheets` etc.

I know this is fixed in dev but it seems like a pretty basic thing so probably worth putting out a release sooner rather than later...

**Comment by jennybc** [comment]
@kleinschmidt I imagine what you're experiencing is running code seen in "dev docs" (the current README and pkgdown site) against the CRAN version. If you install the dev version from GitHub, you will probably be happier..

**Comment by kleinschmidt** [comment]
Ah yeah that's right (I assumed the pkgdown site was built from the last release). I did end up installing master and it works like a charm. Thanks for such a useful package!

**Comment by peterparkerspicklepatch** [comment]
I've run into the problems described above, it's frustrating as the main function of the package is to turn worksheet into a data.frame. Upgrading to dev has been problematic due to building out the dependencies from source. Is there any ETA to push dev to prod?

**Comment by jennybc** [comment]
My main release goal is to make at least one big release before March, with writing fully integrated. Unfortunately, the length of time and number of iterations at CRAN is a sufficient source of friction that I minimize the number of releases to reduce exposure to that.

What dependencies do you have to build from source?

**Comment by peterparkerspicklepatch** [comment]
@jennybc Understandable, I only ask because the package is so useful! I would add documentation that resembles most real world use cases: importing a sheet and then converting to df. I found the documentation hard to follow since it always is pulling the "example" sheet. 

The below is what I think many users are doing:
```
sheet_df <-
  sheets_get("") %>%
  read_sheet(range = "worksheet_name")

And here's it working with the prod version:
```
url <- "url of google sheet here"

sheet_df <-
  read_sheet(
    url,
    range = "worksheet_name" 
    )
```
Since I'm so trained in dplyr, I chain `sheets_get()`

`rlang` is always the dependencies that kills me since it needs other things, then I'm down a rabbit hole.

**Comment by jennybc** [comment]
> rlang is always the dependencies that kills me since it needs other things, then I'm down a rabbit hole.

Understood. But even the development version of googlesheets4 works with the CRAN version of rlang. Using the dev version of googlesheets4 -- a pure R package, no compiled code -- should not require the installation of any other package from source (except gargle, at the moment, which is also pure R).

I have to stage the release of gargle and googlesheets4 leading up to the big March even where googlesheets (the original) will go offline.

>  I found the documentation hard to follow since it always is pulling the "example" sheet.

There is a big section on how to identify and access your own Sheet in googlesheets4's landing page:

https://googlesheets4.tidyverse.org/#identify-and-access-your-own-sheet

**Comment by ivanovskia1** [comment]
Hi @jennybc , I have an existing sheet that I'm trying to append data to 
My google sheet is Sheet1
The dataframe I'm trying to append is df

df %>% sheet_append(Sheet1)

Error: Don't know how to coerce an object of class 'data.frame' into a 'sheets_id'

Any help on this ?

**Comment by jennybc** [comment]
The first argument of `sheet_append()` had to be something that identified a Google Sheet. Above, you seem to be providing the data frame stored as `df`.

https://googlesheets4.tidyverse.org/reference/sheet_append.html
